src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -389,41 +315,210 @@ private static <T extends Enum<T>> int mappedEnumLookup(T value, MethodHandles.L
 
-                resolved = label.resolveConstantDesc(lookup);
-            } catch (IllegalArgumentException | ReflectiveOperationException ex) {
-                resolved = SENTINEL;
-            }
+        private final MethodHandles.Lookup lookup;
+        private final EnumDesc<?>[] enumDescs;
+        @Stable
+        private Object[] resolvedEnum;
 
-            resolvedEnum.resolvedEnum = resolved;
+        public ResolvedEnumLabels(MethodHandles.Lookup lookup, EnumDesc<?>[] enumDescs) {
+            this.lookup = lookup;
+            this.enumDescs = enumDescs;
+            this.resolvedEnum = new Object[enumDescs.length];
         }
 
-        return value == resolvedEnum.resolvedEnum;
-    }
+        @Override
+        public Object apply(Integer labelIndex, Object value) {
+            Object result = resolvedEnum[labelIndex];
 
-    private static MethodHandle withIndexCheck(MethodHandle target, int labelsCount) {
-        MethodHandle checkIndex = MethodHandles.insertArguments(CHECK_INDEX, 1, labelsCount + 1);
+            if (result == null) {
+                try {
+                    EnumDesc<?> label = enumDescs[labelIndex];
+                    Class<?> clazz = label.constantType().resolveConstantDesc(lookup);
 
-        return MethodHandles.filterArguments(target, 1, checkIndex);
-    }
+                    if (value.getClass() != clazz) {
+                        return SENTINEL;
+                    }
 
-    private static final class ResolvedEnumLabel {
-        @Stable
-        public Object resolvedEnum;
+                    result = label.resolveConstantDesc(lookup);
+                } catch (IllegalArgumentException | ReflectiveOperationException ex) {
+                    result = SENTINEL;
+                }
+
+                resolvedEnum[labelIndex] = result;
+            }
+
+            return result;
+        }
     }
 
     private static final class EnumMap {
         @Stable
         public int[] map;
     }
+
+    /*
+     * Construct test chains for labels inside switch, to handle switch repeats:
+     * switch (idx) {
+     *     case 0 -> if (selector matches label[0]) return 0;
+     *     case 1 -> if (selector matches label[1]) return 1;
+     *     ...
+     * }
+     */
+    @SuppressWarnings("removal")
+    private static MethodHandle generateInnerClass(MethodHandles.Lookup caller, Object[] labels) {
+        List<EnumDesc<?>> enumDescs = new ArrayList<>();
+
+        byte[] classBytes = Classfile.of().build(ClassDesc.of(typeSwitchClassName(caller.lookupClass())), clb -> {
+            clb.withFlags(AccessFlag.FINAL, AccessFlag.SYNTHETIC)
+              .withMethod("typeSwitch", MethodTypeDesc.ofDescriptor("(Ljava/lang/Object;ILjava/util/function/BiFunction;)I"), Classfile.ACC_STATIC, mb -> {
+                  mb.withFlags(AccessFlag.PUBLIC, AccessFlag.FINAL, AccessFlag.STATIC)
+                    .withCode(cb -> {
```

Comment:
Can just use `clb.withMethodBody` and pass the access flags with `Classfile.ACC_PUBLIC | Classfile.ACC_STATIC | Classfile.ACC_FINAL`

Reply:
Done.